File: java/dtos/Employee.java
package dtos;

import java.util.*;

public class Employee {
    private String name;
    private Map<DayOfWeek, List<Shift>> preferences; // Stores preferred shifts for each day
    private Map<DayOfWeek, Shift> assignedShifts;    // Stores actual assigned shifts
    private int daysWorked;
    
    public Employee(String name) {
        this.name = name;
        this.preferences = new HashMap<>();
        this.assignedShifts = new HashMap<>();
        this.daysWorked = 0;
        
        // Initialize preferences for all days
        for (DayOfWeek day : DayOfWeek.values()) {
            preferences.put(day, new ArrayList<>());
        }
    }
    
    public String getName() {
        return name;
    }
    
    // Add a preferred shift for a specific day
    public void addPreference(DayOfWeek day, Shift shift) {
        preferences.get(day).add(shift);
    }
    
    // Get preferred shifts for a specific day
    public List<Shift> getPreferences(DayOfWeek day) {
        return new ArrayList<>(preferences.get(day));
    }
    
    // Check if employee has any preferences for a day
    public boolean hasPreferences(DayOfWeek day) {
        return !preferences.get(day).isEmpty();
    }
    
    // Assign a shift to the employee for a specific day
    public void assignShift(DayOfWeek day, Shift shift) {
        if (!isAssigned(day)) {
            assignedShifts.put(day, shift);
            daysWorked++;
        }
    }
    
    // Check if employee is already assigned on a specific day
    public boolean isAssigned(DayOfWeek day) {
        return assignedShifts.containsKey(day);
    }
    
    // Get assigned shift for a specific day
    public Shift getAssignedShift(DayOfWeek day) {
        return assignedShifts.get(day);
    }
    
    // Get number of days worked
    public int getDaysWorked() {
        return daysWorked;
    }
    
    // Check if employee can work more days (max 5 days per week)
    public boolean canWorkMoreDays() {
        return daysWorked < 5;
    }
    
    // Check if employee is available for a specific day
    public boolean isAvailable(DayOfWeek day) {
        return !isAssigned(day) && canWorkMoreDays();
    }
    
    // Clear all assignments (useful for regenerating schedule)
    public void clearAssignments() {
        assignedShifts.clear();
        daysWorked = 0;
    }
    
    // Get all assigned days
    public Set<DayOfWeek> getAssignedDays() {
        return new HashSet<>(assignedShifts.keySet());
    }
    
    @Override
    public String toString() {
        return name;
    }
    
    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Employee employee = (Employee) obj;
        return Objects.equals(name, employee.name);
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(name);
    }
} 

File: python/dto/employee.py
from day_of_week import DayOfWeek
from shift import Shift

class Employee:
    def __init__(self, name):
        self.name = name
        self.preferences = {}  # Stores preferred shifts for each day
        self.assigned_shifts = {}  # Stores actual assigned shifts
        self.days_worked = 0
        
        # Initialize preferences for all days
        for day in DayOfWeek.values():
            self.preferences[day] = []
    
    def get_name(self):
        return self.name
    
    # Add a preferred shift for a specific day
    def add_preference(self, day, shift):
        self.preferences[day].append(shift)
    
    # Get preferred shifts for a specific day
    def get_preferences(self, day):
        return self.preferences[day].copy()
    
    # Check if employee has any preferences for a day
    def has_preferences(self, day):
        return len(self.preferences[day]) > 0
    
    # Assign a shift to the employee for a specific day
    def assign_shift(self, day, shift):
        if not self.is_assigned(day):
            self.assigned_shifts[day] = shift
            self.days_worked += 1
    
    # Check if employee is already assigned on a specific day
    def is_assigned(self, day):
        return day in self.assigned_shifts
    
    # Get assigned shift for a specific day
    def get_assigned_shift(self, day):
        return self.assigned_shifts.get(day)
    
    # Get number of days worked
    def get_days_worked(self):
        return self.days_worked
    
    # Check if employee can work more days (max 5 days per week)
    def can_work_more_days(self):
        return self.days_worked < 5
    
    # Check if employee is available for a specific day
    def is_available(self, day):
        return not self.is_assigned(day) and self.can_work_more_days()
    
    # Clear all assignments (useful for regenerating schedule)
    def clear_assignments(self):
        self.assigned_shifts.clear()
        self.days_worked = 0
    
    # Get all assigned days
    def get_assigned_days(self):
        return set(self.assigned_shifts.keys())
    
    def __str__(self):
        return self.name
    
    def __eq__(self, other):
        if not isinstance(other, Employee):
            return False
        return self.name == other.name
    
    def __hash__(self):
        return hash(self.name) 